p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,508 @@
+
+type iceTimeouts struct {
+	Disconnect time.Duration
+	Failed     time.Duration
+	Keepalive  time.Duration
+}
+
+// WithListenerMaxInFlightConnections sets the maximum number of connections that are in-flight, i.e
+// they are being negotiated, or are waiting to be accepted.
+func WithListenerMaxInFlightConnections(m uint32) Option {
+	return func(t *WebRTCTransport) error {
+		if m == 0 {
+			t.maxInFlightConnections = DefaultMaxInFlightConnections
+		} else {
+			t.maxInFlightConnections = m
+		}
+		return nil
+	}
+}
+
+func New(privKey ic.PrivKey, psk pnet.PSK, gater connmgr.ConnectionGater, rcmgr network.ResourceManager, opts ...Option) (*WebRTCTransport, error) {
+	if psk != nil {
+		log.Error("WebRTC doesn't support private networks yet.")
+		return nil, fmt.Errorf("WebRTC doesn't support private networks yet")
+	}
+	localPeerID, err := peer.IDFromPrivateKey(privKey)
+	if err != nil {
+		return nil, fmt.Errorf("get local peer ID: %w", err)
+	}
+	// We use elliptic P-256 since it is widely supported by browsers.
+	//
+	// Implementation note: Testing with the browser,
+	// it seems like Chromium only supports ECDSA P-256 or RSA key signatures in the webrtc TLS certificate.
+	// We tried using P-228 and P-384 which caused the DTLS handshake to fail with Illegal Parameter
+	//
+	// Please refer to this is a list of suggested algorithms for the WebCrypto API.
+	// The algorithm for generating a certificate for an RTCPeerConnection
+	// must adhere to the WebCrpyto API. From my observation,
+	// RSA and ECDSA P-256 is supported on almost all browsers.
+	// Ed25519 is not present on the list.
+	pk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("generate key for cert: %w", err)
+	}
+	cert, err := webrtc.GenerateCertificate(pk)
+	if err != nil {
+		return nil, fmt.Errorf("generate certificate: %w", err)
+	}
+	config := webrtc.Configuration{
+		Certificates: []webrtc.Certificate{*cert},
+	}
+	noiseTpt, err := noise.New(noise.ID, privKey, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create noise transport: %w", err)
+	}
+	transport := &WebRTCTransport{
+		rcmgr:        rcmgr,
+		gater:        gater,
+		webrtcConfig: config,
+		privKey:      privKey,
+		noiseTpt:     noiseTpt,
+		localPeerId:  localPeerID,
+
+		peerConnectionTimeouts: iceTimeouts{
+			Disconnect: DefaultDisconnectedTimeout,
+			Failed:     DefaultFailedTimeout,
+			Keepalive:  DefaultKeepaliveTimeout,
+		},
+
+		maxInFlightConnections: DefaultMaxInFlightConnections,
+	}
+	for _, opt := range opts {
+		if err := opt(transport); err != nil {
+			return nil, err
+		}
+	}
+	return transport, nil
+}
+
+func (t *WebRTCTransport) Protocols() []int {
+	return []int{ma.P_WEBRTC}
+}
+
+func (t *WebRTCTransport) Proxy() bool {
+	return false
+}
+
+func (t *WebRTCTransport) CanDial(addr ma.Multiaddr) bool {
+	return dialMatcher.Matches(addr)
+}
+
+var webRTCMultiAddr = ma.StringCast("/webrtc")
```

Comment:
The `webrtc` transport has been renamed to `webrtc-direct`.

Reply:
Thanks for the notice. Will modify it! :)

Reply:
Resolved in https://github.com/libp2p/go-libp2p/pull/1999/commits/3b28745667346091c50fcb9077f4e05b15582962.